### PR TITLE
Docker CPU environment

### PR DESCRIPTION
**Pull request by Wiezzel**

### CODE REVIEW

File: golem/envs/__init__.py [hunk 1]
```diff
@@ -18,8 +16,41 @@
...
+
+class EnvConfig(Serializable, ABC):
+    """ Environment-wide configuration. Specifies e.g. available resources. """
+
+
+class Prerequisites(Serializable, ABC):
+    """
+    Environment-specific requirements for computing a task. Distributed with the
+    task header. Providers are expected to prepare (download, install, etc.)
+    prerequisites in advance no to waste computation time.
```
Comment (mfranciszkiewicz) [inline golem/envs/__init__.py]:
```suggestion
    prerequisites in advance not to waste computation time.
```

File: golem/envs/__init__.py [hunk 2]
```diff
@@ -18,8 +16,41 @@
...
+
+
+class Payload(Serializable, ABC):
+    """
+    A definition for Runtime. Environment-specific description of computation to
+    be run. Received when provider is assigned a subtask.
+    """
+
+
+class EnvSupportStatus(NamedTuple):
```
Comment (mfranciszkiewicz) [inline golem/envs/__init__.py:40]: This already exists, see `golem.environments.environment.SupportStatus`

File: golem/envs/docker/__init__.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    """ This exists because NamedTuple must be single superclass """
+
+    def to_dict(self) -> Dict[str, Any]:
+        dict_ = self._asdict()
+        dict_['source'] = str(dict_['source'])
+        return dict_
+
+    @classmethod
+    def from_dict(cls, dict_: Dict[str, Any]) -> 'DockerBind':
+        source = Path(dict_.pop('source'))
```
Comment (mfranciszkiewicz) [inline golem/envs/docker/__init__.py]: Do we really need to modify `dict_`?

File: golem/envs/docker/cpu.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+        logger.info('Environment supported.')
+        return EnvSupportStatus(True)
+
+    @classmethod
+    def _check_docker_version(cls) -> bool:
+        logger.info('Checking docker version...')
+        try:
+            version_string = DockerCommandHandler.run("version")
+        except SubprocessError:
+            logger.info('Checking docker version failed.')
```
Comment (mfranciszkiewicz) [inline golem/envs/docker/cpu.py]: `logger.[error|exception]`?

File: golem/docker/hypervisor/__init__.py [hunk 5]
```diff
@@ -32,6 +32,7 @@ def __init__(self,
 
     @classmethod
     def is_available(cls) -> bool:
+        # FIXME: Implement an actual check
```
Comment (mfranciszkiewicz) [inline golem/docker/hypervisor/__init__.py]: Maybe `raise NotImplementedError` instead?

File: golem/envs/__init__.py [hunk 6]
```diff
@@ -75,78 +118,93 @@ class EnvStatus(Enum):
...
     @abstractmethod
-    def metadata(self) -> EnvMetadata:
+    def parse_prerequisites(cls, prerequisites_dict: Dict[str, Any]) \
+            -> Prerequisites:
+        """ Build Prerequisites struct from supplied dictionary. Returned value
+            is of appropriate type for calling prepare_prerequisites(). """
+        raise NotImplementedError
+
+    @abstractmethod
+    def prepare_prerequisites(self, prerequisites: Prerequisites) -> Deferred:
```
Comment (mfranciszkiewicz) [inline golem/envs/__init__.py]: IMO the name `prepare_prerequisites` doesn't suggest that we may modify some global state here

File: golem/envs/__init__.py [hunk 7]
```diff
@@ -75,78 +118,93 @@ class EnvStatus(Enum):
...
+    @classmethod
+    @abstractmethod
+    def metadata(cls) -> EnvMetadata:
+        """ Get Environment metadata. """
+        raise NotImplementedError
+
+    @classmethod
     @abstractmethod
-    def metadata(self) -> EnvMetadata:
+    def parse_prerequisites(cls, prerequisites_dict: Dict[str, Any]) \
```
Comment (mfranciszkiewicz) [inline golem/envs/__init__.py:150]: Could this be a part of `Prerequisites`?

File: golem/envs/__init__.py [hunk 8]
```diff
@@ -18,8 +16,41 @@
 RuntimeEvent = Any  # TODO: Define runtime events
 
 
-class Payload:
-    pass
+class Serializable(ABC):
```
Comment (mfranciszkiewicz) [inline golem/envs/__init__.py]: Already exists: `golem.core.simpleserializer.DictSerializable`

File: golem/envs/docker/cpu.py (hunk 4 above)
Comment (Wiezzel) [reply golem/envs/docker/cpu.py]: Yep, this should be `exception()`

File: golem/envs/docker/__init__.py (hunk 3 above)
Comment (Wiezzel) [reply golem/envs/docker/__init__.py]: I could make a copy.

File: golem/envs/__init__.py (hunk 2 above)
Comment (Wiezzel) [reply golem/envs/__init__.py:40]: Well, there are two reasons why I'm reluctant to use it:
1. The `_desc` field of `SupportStatus` is kind of oddly coupled with `UnsupportReason` which doesn't fit the new environments picture (e.g. `CONCENT_REQUIRED` or `MASK_MISMATCH` have nothing to do with environments).
2. The current `environments` package is going away when we're done with refactoring, isn't it?

File: golem/docker/hypervisor/__init__.py (hunk 5 above)
Comment (Wiezzel) [reply golem/docker/hypervisor/__init__.py]: OK, why not.

File: golem/envs/__init__.py (hunk 6 above)
Comment (Wiezzel) [reply golem/envs/__init__.py]: I'd gladly welcome some better name suggestion.
`download_prerequisites`?
`install_prerequisites`?

File: golem/envs/__init__.py (hunk 7 above)
Comment (Wiezzel) [reply golem/envs/__init__.py:150]: Not really. The assumption is that caller of this method doesn't have to know the exact type of `Prerequisites` returned. I imagine something like this:

```python
...
env_id = task_header.env_id
prereqs_dict = task_header.prerequisites
env = env_manager.environment(env_id)
prereqs = env.parse_prerequisites(prereqs_dict)
env.prepare_prerequisites(prereqs)
```

File: golem/docker/hypervisor/__init__.py (hunk 5 above)
Comment (Wiezzel) [reply golem/docker/hypervisor/__init__.py]: Actually, this comment is referring to VirtualBox. We need an availability check for `VirtualBoxHypervisor`.

Comment (Wiezzel) [reply golem/docker/hypervisor/__init__.py]: And for Xhyve as well, as it seems.

File: golem/envs/__init__.py (hunk 6 above)
Comment (mfranciszkiewicz) [reply golem/envs/__init__.py]: Perhaps `install`?

Comment (Wiezzel) [reply golem/envs/__init__.py]: OK, will change.

File: golem/envs/docker/cpu.py [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+        return DockerPayload.from_dict(payload_dict)
+
+    def runtime(self, payload: Payload, config: Optional[EnvConfig] = None) \
+            -> DockerCPURuntime:
+        assert isinstance(payload, DockerPayload)
+        if config is not None:
+            assert isinstance(config, DockerCPUConfig)
+        else:
+            config = self.config()
+        logger.info("Creating runtime...")
```
Comment (maaktweluit) [inline golem/envs/docker/cpu.py:288]: 3 dots at the end of a log messages makes me assume it also reports when its done

**Review by maaktweluit — APPROVED** [review]
LGTM!

Nice and clear comments!

File: golem/envs/__init__.py (hunk 2 above)
Comment (mfranciszkiewicz) [reply golem/envs/__init__.py:40]: Yes, we're going to remove the old `environments` package

**Review by mfranciszkiewicz — APPROVED** [review]

**Pull request merged**